### PR TITLE
[FLINK-16587][network] Provide the method for getting unconsumed buffer from RecordDeserializer

### PR DESCRIPTION
**Pull request by pnowojski**
This is a proposed fixup commit (last commit in the this PR) to an existing PR: https://github.com/apache/flink/pull/11507

It's not intended to be merged independently, but squashed with one commit from the original PR before merging the original PR.

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializationTest.java [hunk 1]
```diff
@@ -179,6 +185,77 @@ private static void testSerializationRoundTrip(
...
+	@Test
+	public void testSpanningRecordUnconsumedBuffer() throws Exception {
+		RecordSerializer<SerializationTestType> serializer = new SpanningRecordSerializer<>();
+		RecordDeserializer<SerializationTestType> deserializer =
+			new SpillingAdaptiveSpanningRecordDeserializer<>(
+				new String[]{tempFolder.getRoot().getAbsolutePath()});
+
+		testUnconsumedBuffer(serializer, deserializer, Util.randomRecord(SerializationTestTypeFactory.INT), 1);
+	}
+
```
Comment (AHeise) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializationTest.java:210]: Add test for records with >128 bytes length.

```

	@Test
	public void testLargeSpanningRecordUnconsumedBuffer() throws Exception {
		RecordSerializer<SerializationTestType> serializer = new SpanningRecordSerializer<>();
		RecordDeserializer<SerializationTestType> deserializer =
			new SpillingAdaptiveSpanningRecordDeserializer<>(
				new String[]{tempFolder.getRoot().getAbsolutePath()});


		testUnconsumedBuffer(serializer, deserializer, Util.randomRecord(SerializationTestTypeFactory.BYTE_ARRAY), 1);
	}
```

**Comment by pnowojski** [comment]
I applied your changes @AHeise . I also have found & fixed another bug. Take a look at the new `testLargeSpanningRecordUnconsumedBufferWithLeftOverBytes` test case and modified `org.apache.flink.runtime.io.network.api.serialization.SpillingAdaptiveSpanningRecordDeserializer.SpanningWrapper#getUnconsumedSegment`
```
} else if (recordLength != -1) {
				int leftOverSize = leftOverLimit - leftOverStart;
				int unconsumedSize = Integer.BYTES + accumulatedRecordBytes + leftOverSize;
				DataOutputSerializer serializer = new DataOutputSerializer(unconsumedSize);
				serializer.writeInt(recordLength);
				serializer.write(buffer, 0, accumulatedRecordBytes);
				if (leftOverSize > 0) {
					serializer.write(checkNotNull(leftOverData), leftOverStart, leftOverSize);
				}
				MemorySegment segment = MemorySegmentFactory.allocateUnpooledSegment(unconsumedSize);
				segment.put(0, serializer.getSharedBuffer(), 0, segment.size());
				return Optional.of(segment);
			}
``` 
branch

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java [hunk 2]
```diff
@@ -95,13 +96,14 @@ public Buffer getCurrentBuffer () {
...
+		Optional<MemorySegment> target;
 		if (nonSpanningWrapper.remaining() > 0) {
-			return Optional.of(new NetworkBuffer(nonSpanningWrapper.copyToTargetSegment(), FreeingBufferRecycler.INSTANCE));
+			target = nonSpanningWrapper.getUnconsumedSegment();
 		} else {
-			MemorySegment target = spanningWrapper.copyToTargetSegment();
-			return target != null ? Optional.of(new NetworkBuffer(target, FreeingBufferRecycler.INSTANCE)) : Optional.empty();
+			target = spanningWrapper.getUnconsumedSegment();
 		}
+		return target.map(memorySegment -> new NetworkBuffer(memorySegment, FreeingBufferRecycler.INSTANCE, true, memorySegment.size()));
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java:106]: Actually thought about doing the same thing ;)

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializationTest.java [hunk 3]
```diff
@@ -179,6 +187,126 @@ private static void testSerializationRoundTrip(
...
+		testUnconsumedBuffer(
+			serializer,
+			deserializer,
+			Util.randomRecord(SerializationTestTypeFactory.BYTE_ARRAY),
+			1,
+			new byte[] {42, 43, 44});
+	}
+
+	public void testUnconsumedBuffer(
```
Comment (AHeise) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializationTest.java:238]: nit: overload could be avoided by using `byte... leftOverBytes`

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java [hunk 4]
```diff
@@ -579,27 +584,36 @@ private void addNextChunkFromMemorySegment(MemorySegment segment, int offset, in
...
 					"records.");
 			} else if (recordLength != -1) {
-				return MemorySegmentFactory.wrap(buffer);
+				int leftOverSize = leftOverLimit - leftOverStart;
+				int unconsumedSize = Integer.BYTES + accumulatedRecordBytes + leftOverSize;
+				DataOutputSerializer serializer = new DataOutputSerializer(unconsumedSize);
+				serializer.writeInt(recordLength);
+				serializer.write(buffer, 0, accumulatedRecordBytes);
+				if (leftOverSize > 0) {
+					serializer.write(checkNotNull(leftOverData), leftOverStart, leftOverSize);
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java:608]: nit: `checkNotNull` is not doing anything. The same NPE will be thrown by `serializer#write`. Either leave out or replace by `checkState` with a descriptive messsage.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java (hunk 2 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java:104]: :)

**Comment by pnowojski** [comment]
@flinkbot run travis
@flinkbot run azure

**Comment by pnowojski** [comment]
Incorporated into https://github.com/apache/flink/pull/11507

**Pull request closed without merging**